zebROS_ws/src/behaviors/src/2023_fourber_server.cpp

```diff
@@ -0,0 +1,383 @@
+    load_param_helper(nh_, "base_away_us_cone/middle_node", res, 0.7);
+    load_param_helper(nh_, "base_away_us_cone/middle_node_below", res_bool, false);
+    game_piece_lookup_[fourber_ns::BASE_AWAY_US_CONE][fourber_ns::MIDDLE_NODE] = {res, res_bool};
+    load_param_helper(nh_, "base_away_us_cone/high_node", res, 1.0);
+    load_param_helper(nh_, "base_away_us_cone/high_node_below", res_bool, false);
+    game_piece_lookup_[fourber_ns::BASE_AWAY_US_CONE][fourber_ns::HIGH_NODE] = {res, res_bool};
+
+    load_param_helper(nh_, "safety_high/distance", res, 0.4);
+    safety_high_distance_ = res;
+    load_param_helper(nh_, "safety_high/below", res_bool, false);
+    safety_high_below_ = res_bool;
+
+    load_param_helper(nh_, "safety_intake/distance", res, 0.4);
+    safety_low_distance_ = res;
+    load_param_helper(nh_, "safety_intake/below", res_bool, true);
+    safety_low_below_ = res_bool;
+
+    ddr_.publishServicesTopics();
+    as_.start();
+    FourberINFO("Started Fourber Action server");
+  }
+
+  ~FourberAction2023(void)
+  {
+  }
+
+  // blocks until fourbar is at the correct position
+  bool waitForFourbar(double position) {
+    ros::Rate r = ros::Rate(10);
+    
+    while (true) {
+      ros::spinOnce();
+      FourberINFO("Waiting for fourbar");
+
+      // essentially just keep fourbar where it is now
+			if (as_.isPreemptRequested() || !ros::ok()) {
+        FourberWARN("Fourber Preemepted");
+        controllers_2023_msgs::FourBarSrv last_req;
+        last_req.request.position = fourbar_cur_setpoint_;
+        fourbar_srv_.call(last_req);
+        return false;
+			}
+
+      if (fabs(fourbar_cur_setpoint_-position) <= position_tolerance_) {
+        FourberINFO("Elevator reached position!");
+        break;
+      }
+      r.sleep();
+    }
+
+    return true;
+  }
+
+  // min distance is the minimum distance the forbar must be extended to not cause problems
+  void safetyBoundsAndCallService(double min_distance, bool below) {
+    controllers_2023_msgs::FourBarSrv safety_req;
+
+    // wanting to go to safe position and already at a safe position
+    if (fourbar_cur_setpoint_ >= min_distance && fourbar_cur_position_ >= min_distance) {
+      // nothing to do
+      return; 
+    }
+    // we are trying to go to an illegal position, 
+    else if (fourbar_cur_setpoint_ < min_distance) {
+      safety_req.request.position = min_distance;
+    }
+    // currentlly in illegal position but want to go to legal position
+    else if (fourbar_cur_setpoint_ >= min_distance && fourbar_cur_position_ < min_distance) {
+      safety_req.request.position = fourbar_cur_setpoint_;
+    }
+
+    safety_req.request.below = below;
+    fourbar_srv_.call(safety_req);
+    if (!waitForFourbar(safety_req.request.position)) {
+      FourberERR("Failed calling fourber with message 'safety_req'");
+      publishFailure();
+    }
+    else {
+      publishSuccess();
+    }
+  }
+
+  void publishFailure() {
+    feedback_.success = false;
```

Comment:
I like the idea of these functions, but feedback_ and result_ could probably be made local instead of member vars?